### ISSUE TITLE
ModifiedFile.source_code invalid for binary files.

### ISSUE THREAD

**Issue by rick-c-4499** [ishepard/pydriller#215]
When a ModifiedFile object is created for a binary file, the source_code field contains string data that was decoded incorrectly and thus is unusable.  This problem stems from _get_decoded_str not properly identifying the encoding of the blob prior to decoding it as utf-8 and ignoring errors. There no way to encode or in any way retrieve the actual source_code (binary) as the data is clobbered by the decode.

Expected results: there should be a way to retrieve the binary data of a ModifiedFile. Either a separate property (maybe source_code_bytes) or the type of source_code could vary.

**Comment by ishepard** [comment]
Hi Rick,
yeah indeed ModifiedFile does not check the type of the file. I think this is kind of excepted, we can not decode a binary. 

> there should be a way to retrieve the binary data of a ModifiedFile

What do you mean exactly by this?  retrieving binary data?

>  Either a separate property (maybe source_code_bytes) or the type of source_code could vary.

Maybe, but then we need to have a logic to identify binary files.

What I saw people doing is that they filter for files they are interested in, e.g.:
```
for mod_file in modified_files:
    if mod_file.filename.endswith('.java'):
        ....
```

**Comment by rick-c-4499** [comment]
>  I think this is kind of excepted [expected?], we can not decode a binary.

I'm not expecting the library to decode a binary. I was expecting the library to not ATTEMPT to decode it when it cannot do so. I do agree that my overloading the expectations of what ModifiedFile.source_code should be used for regarding binary files IS a bit of a stretch but it's a bit of a trap to exist and be populated with incorrectly decoded data. For example other fields like ModifiedFile.diff correctly handle binary files by containing a string which states: ''Binary files % and % differ" in lieu of actual source code diffs which are impossible (I understand that's basically passing on raw git behavior but my point stands).

My use case is I'm searching through a repo to find Word Macro files (.dotm) which are binary. I'll then use another library to decode the contents for further processing.  As mentioned above I was expecting to use .source_code to access the contents of the binary files (to decode another way) but was surprised to find the contents was populated but corrupted due to the pre-emptive decode call which was making an assumption that is not always valid. I worked around this by dropping to the gitPython and getting the git.Blob.data_stream but it would be nice to stay in pydriller.

> Maybe, but then we need to have a logic to identify binary files.

I'm not super familiar with git but surly there is a trivial way to detect that. Again I feel like just assuming the git blob is always decodable as utf-8 is incorrect.

Overall I'm happy to offer up a pull request but would like to defer to you as to the API design. Should the type (str vs bytes or something) of ModifiedFile.source_code vary depending on the underlying blob contents, should there be a niche ModifiedFile.source_code_bytes property, or something else entirely?

**Comment by rick-c-4499** [comment]
Thanks for your offering of workaround suggestions, however I'm good on that front.

I solely want to point out the library as is has unexpected behavior when it comes to using Modifiedfile.source_code when the file itself is binary. If my expectations are off then I'm happy to close this as out of scope.

**Comment by ishepard** [comment]
@rick-c-4499 no I actually agree with you. 
We currently decode everything, no matter what, even files we shouldn't decode.
A smarter approach would be to only decode when possible.

It's the "when possible" that's tricky. We can't differentiate by file, because ANY file can have a non UTF-8 character.

So maybe, when we hit an error in decoding, we should simply return the string, without decoding it. Then the user will do whatever they want. 
I think this should be possible.

**Comment by stefanodallapalma** [comment]
Personally, I believe the code semantic should stick to the method/property name. I would not expect `ModifiedFile.source_code` to return the content of a binary file, either encoded or decoded, but that's my point of view of course :)

Would it be better to add a property like `ModifiedFile.content` that returns the content no matter what so to let the user do whatever they want, as @ishepard suggested? 

As for the property `source_code`, an idea would be returning the value only for files that have an extension supported by known programming languages (.kt for Kotlin, .py for Python) but I am afraid it would be difficult to maintain.

**Comment by rick-c-4499** [comment]
I do agree with you that having non decoded "binary" sometimes in source_code breaks the semantic API contract.

I ALSO worry that having "decoded gibberish" in source_code breaks the semantic expectations and further agree that any sort of "smarter" parsing/handling is too costly/impractical.

Perhaps the patch would be:

Change the decode utf8 to not ignore errors. Per @ishepard 
On an error store the empty string or some sort of error string in source_code to match expectations for @stefanodallapalma 
Create a new .source_code_bytes or some other property which always contains the undecoded contents? Though I do realize this is a very niche use and there are workarounds (as mentioned above). Perhaps you could document the empty/error string in source_code and add a code snippet of the workaround such that a property isn't needed.

However I'm also happy enough with source_code breaking the sematic expectations and storing the utf8 decoded string when there are no errors but on the rare occasion that there is an error fall back to just storing the undecoded bytes expecting the user to understand what encoding the file is and decode it themselves.

**Comment by ishepard** [comment]
>  I would not expect ModifiedFile.source_code to return the content of a binary file, **either encoded or decoded,** 
Yeah exactly. Right now, `.source_code` returns the decoded string of a binary file, which is wrong of course. We shouldn't do it. 

> As for the property source_code, an idea would be returning the value only for files that have an extension supported by known programming languages (.kt for Kotlin, .py for Python) but I am afraid it would be difficult to maintain.
Definitely not indeed, too hard.

> Would it be better to add a property like ModifiedFile.content
Yeah I like this idea

@rick-c-4499 to summarize:
- `.source_code` will not ignore errors. When we hit an error, it means that we failed in decoding. Hence, we return an empty string.
- `.content_undecoded` will simply be a `str()` of whatever GitPython passes us.

### Why content_undecoded?
Working on this task just made me realize that we shouldn't call it `ModifiedFile.source_code`, but actually `ModifiedFile.content`. I think I wrote source_code because I mainly analyse source code files, but it should be more generic. I will create a new issue with this. I will check the right terminology with Git and report back.

**Comment by stefanodallapalma** [comment]
@ishepard 

If we are introducing `ModifiedFile.content_undecoded` I'd suggest to simply deprecate `ModifiedFile.source_code` -- i.e., no need to modify it -- and throw a DeprecationWarning suggesting the user to use `ModifiedFile.content_undecoded` instead.
